clash-ghc/src-ghc/Clash/GHC/LoadInterfaceFiles.hs

```diff
@@ -208,29 +208,30 @@ loadExternalExprs'
 loadExternalExprs' _hdl loaded visited [] =
   return (loaded, visited)
 loadExternalExprs' hdl loaded0 visited0 (e:es) = do
-  let fvs = CoreFVs.exprSomeFreeVarsList
-              (\v -> Var.isId v &&
-                     isNothing (Id.isDataConId_maybe v) &&
-                     not (v `UniqSet.elementOfUniqSet` visited0)
-              ) e
+  let
+    isInteresting v =
+         Var.isId v
+      && not (v `UniqSet.elementOfUniqSet` visited0)
+      && isNothing (Id.isDataConId_maybe v)
 
-      (clsOps',fvs') = partition (isJust . Id.isClassOpId_maybe) fvs
+    fvs0 = CoreFVs.exprSomeFreeVarsList isInteresting e
+    fvs1 = map (\v -> maybe (Left v) (Right . (v,)) (Id.isClassOpId_maybe v)) fvs0
+    (fvs2, clsOps0) = partitionEithers fvs1
+    clsOps1 = map goClsOp clsOps0
 
-      clsOps'' = map
-        ( \v -> flip (maybe (error $ $(curLoc) ++ "Not a class op")) (Id.isClassOpId_maybe v) $ \c ->
-            let clsIds = Class.classAllSelIds c
-            in  maybe (error $ $(curLoc) ++ "Index not found")
-                      (v,)
-                      (elemIndex v clsIds)
-        ) clsOps'
-
-  loaded1 <- mergeLoadedBinders <$> mapM (loadExprFromIface hdl) fvs'
+  loaded1 <- mergeLoadedBinders <$> mapM (loadExprFromIface hdl) fvs2
 
   loadExternalExprs'
     hdl
-    (mergeLoadedBinders [loaded0, loaded1, emptyLb{lbClassOps=clsOps''}])
-    (foldl' UniqSet.addListToUniqSet visited0 [collectLbBinders loaded1, clsOps'])
+    (mergeLoadedBinders [loaded0, loaded1, emptyLb{lbClassOps=clsOps1}])
+    (foldl' UniqSet.addListToUniqSet visited0 [collectLbBinders loaded1, map fst clsOps0])
     (es ++ map snd (lbBinders loaded1))
+ where
+  goClsOp :: (Var.Var, GHC.Class) -> (CoreSyn.CoreBndr, Int)
+  goClsOp (v, c) =
+    case elemIndex v (Class.classAllSelIds c) of
+      Nothing -> error ($(curLoc) <> "Internal error: Index not found")
```

Comment:
This refactoring would've been a good chance to improve the error message.

Reply:
I'm not sure what's happening here though, what would be a good error message?

Reply:
Looks like the error happens when a var isn't in a class. Maybe something like _could not find a binder for <var> in <class>_ if you can get nice names for each

Reply:
Yeah, if we error out it means that we couldn't find a class-method called `v` in the class `c`.